qualtran/bloqs/rotations/phase_gradient_test.py

```diff
@@ -77,95 +77,122 @@ def test_phase_gradient_gate(n: int, exponent, controlled):
+                input_state = int(f'{x:0{x_bit}b}' + f'{phase_grad:0{phase_bit}b}', 2)
+                output_state = int(f'{x:0{x_bit}b}' + f'{phase_grad_out:0{phase_bit}b}', 2)
                 expected_unitary[output_state, input_state] = 1
-    # Test cirq style simulation.
-    assert len(basis_map) == len(set(basis_map.values()))
+            else:
+                for control in [0, 1]:
+                    phase_grad_out_ctrld = (
+                        phase_grad_out if controlled_by == control else phase_grad
+                    )
+                    input_state = int(
+                        f'{control}' + f'{x:0{x_bit}b}' + f'{phase_grad:0{phase_bit}b}', 2
+                    )
+                    output_state = int(
+                        f'{control}' + f'{x:0{x_bit}b}' + f'{phase_grad_out_ctrld:0{phase_bit}b}', 2
+                    )
+                    expected_unitary[output_state, input_state] = 1
+
+    # verify bloq unitary
     circuit = cirq.Circuit(bloq.on(*cirq.LineQubit.range(num_bits)))
     np.testing.assert_allclose(circuit.unitary(), expected_unitary, atol=1e-8)
 
 
-@pytest.mark.slow
+def test_add_into_phase_grad_t_complexity():
+    x_bit, phase_bit = 4, 7
+    bloq = AddIntoPhaseGrad(x_bit, phase_bit)
+    ((toffoli, n),) = bloq.bloq_counts().items()
+    assert bloq.t_complexity() == n * toffoli.t_complexity()
+
+
+_ADD_SCALED_VAL_INTO_PHASE_REG_EXAMPLES: list[AddScaledValIntoPhaseReg] = [
+    AddScaledValIntoPhaseReg.from_bitsize(4, 7, 0.123, 6),
+    AddScaledValIntoPhaseReg.from_bitsize(2, 8, 1.3868682, 8),
+    AddScaledValIntoPhaseReg.from_bitsize(4, 9, -9.0949456, 5),
+    AddScaledValIntoPhaseReg.from_bitsize(6, 4, 2.5, 2),
+    AddScaledValIntoPhaseReg(QFxp(4, 0, signed=False), 4, 1.3868682, QFxp(8, 7, signed=False)),
+]
+
+
 @pytest.mark.parametrize(
     'bloq',
     [
-        AddScaledValIntoPhaseReg.from_bitsize(4, 7, 0.123, 6),
-        AddScaledValIntoPhaseReg.from_bitsize(2, 8, 1.3868682, 8),
-        AddScaledValIntoPhaseReg.from_bitsize(4, 9, -19.0949456, 5),
-        AddScaledValIntoPhaseReg.from_bitsize(6, 4, 2.5, 2),
-        AddScaledValIntoPhaseReg(QFxp(4, 0, signed=False), 4, 1.3868682, QFxp(8, 7, signed=False)),
+        pytest.param(bloq, marks=pytest.mark.slow if bloq.num_qubits() > 10 else ())
+        for bloq in _ADD_SCALED_VAL_INTO_PHASE_REG_EXAMPLES
     ],
 )
-def test_add_scaled_val_into_phase_reg(bloq):
+def test_add_scaled_val_into_phase_reg_classical_sim(bloq: AddScaledValIntoPhaseReg):
     cbloq = bloq.decompose_bloq()
     for x in range(2**bloq.x_dtype.bitsize):
         for phase_grad in range(2**bloq.phase_bitsize):
             d = {'x': x, 'phase_grad': phase_grad}
             c1 = bloq.on_classical_vals(**d)
             c2 = cbloq.on_classical_vals(**d)
             assert c1 == c2, f'{d=}, {c1=}, {c2=}'
+
+
+@pytest.mark.parametrize(
+    'bloq',
+    [
+        pytest.param(bloq, marks=pytest.mark.slow if bloq.num_qubits() > 12 else ())
```

Comment:
bloq2 is still getting executed on my machine (7 seconds) despite it looking like it should be 4+9=13 qubits

Reply:
consider using bloq.signature.n_qubits, which is explicit about how it handles left/right registers

Reply:
About bloq2, my guess would be that because `AddIntoPhaseGrad` doesn't define a decomposition, but only a classical action, and cirq.unitary computes the gate much faster for classical gates.